internal/p2p/channel.go

```diff
@@ -0,0 +1,208 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/tendermint/tendermint/types"
+)
+
+// Envelope contains a message with sender/receiver routing info.
+type Envelope struct {
+	From      types.NodeID  // sender (empty if outbound)
+	To        types.NodeID  // receiver (empty if inbound)
+	Broadcast bool          // send to all connected peers (ignores To)
+	Message   proto.Message // message payload
+
+	// channelID is for internal Router use, set on outbound messages to inform
+	// the sendPeer() goroutine which transport channel to use.
+	//
+	// FIXME: If we migrate the Transport API to a byte-oriented multi-stream
+	// API, this will no longer be necessary since each channel will be mapped
+	// onto a stream during channel/peer setup. See:
+	// https://github.com/tendermint/spec/pull/227
+	channelID ChannelID
+}
+
+// Wrapper is a Protobuf message that can contain a variety of inner messages
+// (e.g. via oneof fields). If a Channel's message type implements Wrapper, the
+// Router will automatically wrap outbound messages and unwrap inbound messages,
+// such that reactors do not have to do this themselves.
+type Wrapper interface {
+	proto.Message
+
+	// Wrap will take a message and wrap it in this one if possible.
+	Wrap(proto.Message) error
+
+	// Unwrap will unwrap the inner message contained in this message.
+	Unwrap() (proto.Message, error)
+}
+
+// PeerError is a peer error reported via Channel.Error.
+//
+// FIXME: This currently just disconnects the peer, which is too simplistic.
+// For example, some errors should be logged, some should cause disconnects,
+// and some should ban the peer.
+//
+// FIXME: This should probably be replaced by a more general PeerBehavior
+// concept that can mark good and bad behavior and contributes to peer scoring.
+// It should possibly also allow reactors to request explicit actions, e.g.
+// disconnection or banning, in addition to doing this based on aggregates.
+type PeerError struct {
+	NodeID types.NodeID
+	Err    error
+}
+
+func (pe PeerError) Resolve() error { return fmt.Errorf("peer=%q: %w", pe.NodeID, pe.Err) }
+func (pe PeerError) Error() string  { return pe.Resolve().Error() }
+
+// Channel is a bidirectional channel to exchange Protobuf messages with peers,
+// wrapped in Envelope to specify routing info (i.e. sender/receiver).
```

Comment:
Minor nit: Avert confusion about whether the channel or the message is wrapped:
```suggestion
// Channel is a bidirectional channel to exchange Protobuf messages with peers.
// Each message is wrapped in an Envelope to specify its sender and receiver.
```